### PR TITLE
[KEYCLOAK-405] - Feature that doesn't allow old password to be reused

### PR DESCRIPTION
**Pull request by girirajsharma**
Tested UI manually for now by firing certain values of n, increasing values of n , decreasing values of n, increasing and then decreasing values of n and vice versa.

### CODE REVIEW

File: model/api/src/main/java/org/keycloak/models/PasswordPolicy.java [hunk 1]
```diff
@@ -18,16 +22,17 @@
     public static final String INVALID_PASSWORD_MIN_SPECIAL_CHARS_MESSAGE = "invalidPasswordMinSpecialCharsMessage";
     public static final String INVALID_PASSWORD_NOT_USERNAME = "invalidPasswordNotUsernameMessage";
     public static final String INVALID_PASSWORD_REGEX_PATTERN = "invalidPasswordRegexPatternMessage";
+    public static final String INVALID_EXPIRED_PASSWORD = "invalidExpiredPasswordMessage";
 
-    private List<Policy> policies;
-    private String policyString;
+    private static List<Policy> policies;
```
Comment (stianst) [inline model/api/src/main/java/org/keycloak/models/PasswordPolicy.java]: These should be static!

File: model/api/src/main/java/org/keycloak/models/UserCredentialModel.java [hunk 2]
```diff
@@ -8,6 +8,7 @@
  */
 public class UserCredentialModel {
     public static final String PASSWORD = "password";
+    public static final String PASSWORD_EXPIRED = "password-expired";
```
Comment (stianst) [inline model/api/src/main/java/org/keycloak/models/UserCredentialModel.java]: Do you think "password-history" would be better?

File: model/mongo/src/main/java/org/keycloak/models/mongo/keycloak/adapters/UserAdapter.java [hunk 3]
```diff
@@ -29,7 +33,9 @@
  * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
  */
 public class UserAdapter extends AbstractMongoAdapter<MongoUserEntity> implements UserModel {
-
+    
+    protected static final int MAX_OLD_PASSWORDS_NOT_ALLOWED = 3;
```
Comment (stianst) [inline model/mongo/src/main/java/org/keycloak/models/mongo/keycloak/adapters/UserAdapter.java]: Why is this here?

File: model/mongo/src/main/java/org/keycloak/models/mongo/keycloak/adapters/UserAdapter.java [hunk 4]
```diff
@@ -177,31 +183,73 @@ public void setTotp(boolean totp) {
 
     @Override
     public void updateCredential(UserCredentialModel cred) {
+
+        if (cred.getType().equals(UserCredentialModel.PASSWORD)) {
+            updatePasswordCredential(cred);
```
Comment (stianst) [inline model/mongo/src/main/java/org/keycloak/models/mongo/keycloak/adapters/UserAdapter.java]: Rather than duplicating the updatePasswordCredential method in all stores we should rename CredentialValidation to CredentialUtils and add it as a static method there

**Comment by stianst** [comment]
Added a few comments. One major issue here is fields in PasswordPolicy that are made static that'll will introduce some weird concurrency issues, for example one realm could suddenly use another realms policy.

**Comment by girirajsharma** [comment]
@stianst Thanks, I will refactor the changes and will squash the commits.

File: model/api/src/main/java/org/keycloak/models/UserCredentialModel.java (hunk 2 above)
Comment (girirajsharma) [inline model/api/src/main/java/org/keycloak/models/UserCredentialModel.java]: "password-history"  makes more sense here. But, for all other message literals or error literals or UI strings, IMHO, its better to keep term as expired rather than history.
eg.
public static final String INVALID_EXPIRED_PASSWORD = "invalidExpiredPasswordMessage";
